plugins/hotspot.c

```diff
@@ -0,0 +1,253 @@
+  if(ret) {
+    fprintf(stderr, "Hotstat: Couldn't initialize the frequency table on thread %d!\n",
+      mambo_get_thread_id(ctx));
+    exit(1);
+  }
+
+  ret = mambo_set_thread_plugin_data(ctx, (void*) basic_block_freq);
+  if(ret) {
+    fprintf(stderr, "Hotstat: Couldn't set the thread plugin data on thread %d!\n",
+      mambo_get_thread_id(ctx));
+    exit(1);
+  }
+}
+
+int hotstat_post_thread_cb(mambo_context* ctx) {
+  int ret;
+
+  mambo_ht_t* basic_block_freq = (mambo_ht_t*) mambo_get_thread_plugin_data(ctx);
+  if(basic_block_freq == NULL) {
+    fprintf(stderr, "Hotstat: Couldn't get the thread plugin data on thread %d!\n",
+      mambo_get_thread_id(ctx));
+    exit(1);
+  }
+
+  mambo_ht_t* global_basic_block_freq = (mambo_ht_t*) mambo_get_plugin_data(ctx);
+  if(global_basic_block_freq == NULL) {
+    fprintf(stderr, "Hotstat: Couldn't get the plugin data on thread %d!\n",
+      mambo_get_thread_id(ctx));
+    exit(1);
+  }
+
+  __mambo_ht_lock(global_basic_block_freq);
+
+  for(int index = 0; index < basic_block_freq->size; index++) {
+
+    uintptr_t key = basic_block_freq->entries[index].key;
+    uint64_t* counter = (uint64_t*) basic_block_freq->entries[index].value;
+
+    uint64_t* global_counter;
+
+    // Find valid elements in the per thread hash map
+    if(key != 0) {
+      // Check of the key is already in the global hash map
+      if(mambo_ht_get_nolock(global_basic_block_freq, key, (uintptr_t*) &global_counter)) {
+        // Add new entry if the key is not in the global map
+        global_counter = (uint64_t*) mambo_alloc(ctx, sizeof(uint64_t));
+        if(global_counter == NULL) {
+          fprintf(stderr, "Hotstat: Couldn't allocate the global counter on thread %d!\n",
+            mambo_get_thread_id(ctx));
+          exit(1);
+        }
+
+        *global_counter = *counter;
+
+        ret = mambo_ht_add_nolock(global_basic_block_freq, key, (uintptr_t) global_counter);
+        if(ret) {
+          fprintf(stderr, "Hotstat: Couldn't add entry to the global frequency table on thread %d!\n",
+            mambo_get_thread_id(ctx));
+          exit(1);
+        }
+
+        mambo_free(ctx, counter);
+      }
+      else {
+        // Add the count from the local map to the global one if the entry already exists
+        *global_counter = *global_counter + *counter;
+      }
+    }
+  }
+
+  __mambo_ht_unlock(global_basic_block_freq);
+
+  mambo_free(ctx, basic_block_freq);
+}
+
+int hotstat_pre_basic_block_cb(mambo_context* ctx) {
+  int ret;
+
+  void* addr = mambo_get_source_addr(ctx);
+
+  if((uintptr_t) addr >= HOTSTAT_MAX_BB_ADDR)
+    return 0;
+
+  mambo_ht_t* basic_block_freq = (mambo_ht_t*) mambo_get_thread_plugin_data(ctx);
+  if(basic_block_freq == NULL) {
+    fprintf(stderr, "Hotstat: Couldn't get the thread plugin data on thread %d!\n",
+      mambo_get_thread_id(ctx));
+    exit(1);
+  }
+
+  uint64_t* counter = NULL;
+
```

Comment:
We don't need separate logic depending on the type of fragment. The same basic block in the application may be scanned multiple times (e.g. because of code cache flushing) so it should always use the code path which attempts a lookup first.